File: src/singleton/java/LazySingleton.java
package singleton.java;

/**
 * 懒汉式
 */
public class LazySingleton {

    // volatile 关键字保证当sInstance被初始化，能够及时被其它线程得知
    private volatile static LazySingleton sInstance;

    /**
     * 构造函数私有化
     */
    private LazySingleton() {

    }

    /**
     * 提供静态方法返回实例
     * 使用双重检查锁的方式
     */
    public static LazySingleton getInstance() {
        if (null == sInstance) {
            synchronized (LazySingleton.class) {
                if (sInstance == null) {
                    sInstance = new LazySingleton();
                }
            }
        }
        return sInstance;
    }
}

File: src/singleton/kotlin/LazySingleton.kt
package singleton.kotlin

/**
 * 懒汉式
 * 使用的是object关键字的另一种用法：伴生对象
 */
class LazySingleton private constructor(){

    companion object {
        private var instance: LazySingleton? = null
        get(){
            if (field == null) {
                field = LazySingleton()
            }
            return field
        }

        /**
         * 直接在get方法上加上锁，保证同一个时刻只有一处能够调用。
         */
        @Synchronized
        fun get(): LazySingleton {
            return instance!!
        }
    }
}